src/components/tl/sharp/tl_sharp_context.c

```diff
@@ -0,0 +1,323 @@
+    ucc_tl_sharp_rcache_region_t *region;
+    void *address;
+    size_t length;
+    int ret;
+
+    address = (void*)rregion->super.start;
+    length  = (size_t)(rregion->super.end - rregion->super.start);
+    region  = ucc_derived_of(rregion, ucc_tl_sharp_rcache_region_t);
+
+    ret = sharp_coll_reg_mr(ctx->sharp_context, address,
+                            length, &region->reg.mr);
+    if (ret < 0) {
+        tl_error(ctx->super.super.lib, "reg failed(%d). addr:%p len:%ld",
+                 ret, address, length);
+        return UCS_ERR_INVALID_PARAM;
+    } else {
+        tl_debug(ctx->super.super.lib, "region:%p reg mr:%p addr:%p len:%ld",
+                 rregion, region->reg.mr, address, length);
+        return UCS_OK;
+    }
+}
+
+static void ucc_tl_sharp_rcache_mem_dereg_cb(void *context, ucc_rcache_t *rcache,
+                                             ucc_rcache_region_t *rregion)
+{
+    ucc_tl_sharp_context_t *ctx          = (ucc_tl_sharp_context_t*)context;
+    ucc_tl_sharp_rcache_region_t *region = ucc_derived_of(rregion,
+                                                ucc_tl_sharp_rcache_region_t);
+    int ret;
+
+    ret = sharp_coll_dereg_mr(ctx->sharp_context, region->reg.mr);
+    if (ret < 0) {
+        tl_error(ctx->super.super.lib, "dereg faied(%d). mr:%p",
+                 ret, region->reg.mr);
+    } else {
+        tl_debug(ctx->super.super.lib, "rregion:%p dereg mr:%p",
+                 rregion, region->reg.mr);
+    }
+}
+
+static void
+ucc_tl_sharp_rcache_dump_region_cb(void *context, ucs_rcache_t *rcache,
+                                   ucs_rcache_region_t *rregion, char *buf,
+                                   size_t max)
+{
+    ucc_tl_sharp_rcache_region_t *region = ucc_derived_of(rregion,
+                                           ucc_tl_sharp_rcache_region_t);
+
+    snprintf(buf, max, "bar ptr:%p", region->reg.mr);
+}
+
+static ucc_rcache_ops_t ucc_tl_sharp_rcache_ops = {
+    .mem_reg     = ucc_tl_sharp_rcache_mem_reg_cb,
+    .mem_dereg   = ucc_tl_sharp_rcache_mem_dereg_cb,
+    .dump_region = ucc_tl_sharp_rcache_dump_region_cb
+};
+
+UCC_CLASS_INIT_FUNC(ucc_tl_sharp_context_t,
+                    const ucc_base_context_params_t *params,
+                    const ucc_base_config_t *config)
+{
+    ucc_tl_sharp_context_config_t *tl_sharp_config =
+        ucc_derived_of(config, ucc_tl_sharp_context_config_t);
+    struct sharp_coll_init_spec init_spec = {0};
+    ucc_status_t status;
+    ucc_rcache_params_t rcache_params;
+
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_sharp_config->super.tl_lib,
+                              params->context);
+    memcpy(&self->cfg, tl_sharp_config, sizeof(*tl_sharp_config));
+
+    if (self->cfg.rand_seed == 0) {
+        self->cfg.rand_seed = time(NULL);
+    }
+
+    self->rcache           = NULL;
+    self->oob_ctx.ctx      = self;
+    self->oob_ctx.ctx_oob  = &UCC_TL_CTX_OOB(self);
```

Comment:
Need to actually check if CTX_OOB is there. 
We can check the params->context->ctx_rank != UCC_RANK_MAX. Otherwise need to bail with silent "debug" message.
Alternatively we could report "ctx_oob_required" vial lib->attr. then CORE would not even try to enable sharp whithout oob.

Reply:
added mask flag check